### PR TITLE
49 create login route

### PR DESCRIPTION
**Pull request by Bas-Shiekh**

### CODE REVIEW

**Review by Aaya-Elsharief — APPROVED** [review]

File: server/controllers/auth/login.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    if (!count) {
+      throw new CustomError(401, "email doesn't exist");
+    }
+    const hashedPassword = await compare(loginPassword, rows[0].getDataValue('password'));
+    if (!hashedPassword) {
+      throw new CustomError(401, 'invalid password');
+    }
+    res.json({ mag: 'logged in successfully' });
+  } catch (err) {
+    res.json({ msg: err });
```
Comment (Raghad-Mezied) [inline server/controllers/auth/login.ts:27]: Here just pass the error to the next error middleware and it will be handled there

**Review by Raghad-Mezied — CHANGES_REQUESTED** [review]

File: server/controllers/auth/login.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    await loginValidate({ email, loginPassword });
+    const { count, rows }: any = await loginQuery(email);
+    if (!count) {
+      throw new CustomError(400, "email doesn't exist");
+    }
+    const hashedPassword = await compare(loginPassword, rows[0].getDataValue('password'));
+    if (!hashedPassword) {
+      throw new CustomError(400, 'invalid password');
+    }
+    await generateToken({ id: rows[0].getDataValue('id'), username: rows[0].getDataValue('name'), role: rows[0].getDataValue('role') });
```
Comment (Raghad-Mezied) [inline server/controllers/auth/login.ts:18]: Rename username to name

**Review by Raghad-Mezied — APPROVED** [review]

**Review by Raghad-Mezied — DISMISSED** [review]

File: server/utils/jwt/index.ts [hunk 3]
```diff
@@ -23,7 +23,20 @@ const verifyToken = (token: string) => new Promise((resolve, reject) => {
...
+) => new Promise((res, rej) => {
+  sign(payload, secretKey as Secret, (error, decoded) => {
+    if (error) {
+      rej(error);
+    } else {
+      res(decoded);
+    }
+  });
+});
+
```
Comment (moustf) [inline server/utils/jwt/index.ts:37]: Why did you create the generateToken function, the singToken function does the same purpose!

File: server/routes/auth.ts [hunk 4]
```diff
@@ -1,10 +1,13 @@
...
+  signup, logout, userData, login,
+} from '../controllers';
 import { userAuth } from '../middlewares';
 
 const authRouter = require('express').Router();
 
 authRouter.post('/auth/signup', signup);
 authRouter.post('/auth/logout', logout);
 authRouter.get('/auth/', userAuth, userData);
+authRouter.post('/login', login);
```
Comment (moustf) [inline server/routes/auth.ts:11]: Let the path be '/auth/login' to follow the convention.

**Review by moustf — APPROVED** [review]

**Review by Aaya-Elsharief — DISMISSED** [review]

**Review by Aaya-Elsharief — APPROVED** [review]

**Review by moustf — APPROVED** [review]

**Review by Raghad-Mezied — APPROVED** [review]

**Pull request merged**